xc/domain.ml

```diff
@@ -453,17 +453,22 @@ let destroy (task: Xenops_task.task_handle) ~xc ~xs ~qemu_domid ~dm domid =
     (fun pcidev ->
        let open Xenops_interface.Pci in
        log_exn_continue
-         ("Deassign PCI device " ^ string_of_address pcidev)
-         (fun () ->
-            Xenctrl.domain_deassign_device xc domid
-              (pcidev.domain, pcidev.bus, pcidev.dev, pcidev.fn)) ())
+         ("Reset PCI device " ^ string_of_address pcidev)
+         (fun () -> Device.PCI.reset ~xs pcidev) ())
     all_pci_devices;
+  
+  (* PCI specification document says that the Function must complete the FLR within 100 ms
+    https://pcisig.com/sites/default/files/specification_documents/ECN_RN_29_Aug_2013.pdf on page 7 *)
+  Thread.delay 0.1;
```

Comment:
I think it is better to put the delay into `Device.PCI.reset` itself, to have that function return only when it is confident that the reset is complete. That would fix the problem for other callers of the function as well (even if we may not have any right now).

Reply:
Hi Rob, these reset operations are in synchronous pattern, so if we passthrough many devices to a VM, saying 20. This will add up to 2 seconds.
Also the delay only matters before a de-assigning, as we need to ensure a clean device to be de-assigned, while I suppose there could be some cases where we just want to send a reset signal and not that care about if the reset complete.